refactor(polys): Use mpmath mpf for RR and CC

oscarbenjamin opened the pull request:
Remove the MPElement subclasses of mpmath's mpf and mpc and SymPy's custom MPContext subclass. This means that RR and CC only use mpmath's public interface.

This also fixes a bug where all RealField domains share a global precision.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* polys
    * The RR and CC domains now use mpmath's mpf and mpc types directly rather than making custom subclasses of mpf, mpc and PythonMPContext. This ensures that only mpmath's public interface is used. This also means that elements of RR and CC no longer have a `.parent` attribute (as is also the case for ZZ , QQ and GF(p)). A bug causing all RealField domains to share a global precision setting was also fixed. The tolerance parameter of RealField and ComplexField is also removed.
    * BREAKING CHANGE: The elements of RR and CC no longer have a `.parent()` method. This is because they are no plain mpmath `mpf` and `mpc` instances rather than subclasses.
<!-- END RELEASE NOTES -->

oscarbenjamin commented on sympy/polys/domains/tests/test_domains.py:
```diff
@@ -1281,11 +1281,6 @@ def test_canonical_unit():
     assert (K.one + i)/(i - K.one) == -i
 
 
-def test_issue_18278():
-    assert str(RR(2).parent()) == 'RR'
-    assert str(CC(2).parent()) == 'CC'
```
The issue referred to here is gh-18278 which notes that `RR(2).parent()` previously returned `CC`.

More generally lower-level domain elements don't have a `.parent()` method (unlike in Sage):
```python
>>> from sympy import *
>>> ZZ.one.parent()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'flint.types.fmpz.fmpz' object has no attribute 'parent'
>>> QQ.one.parent()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'flint.types.fmpq.fmpq' object has no attribute 'parent'
>>> GF(2).one.parent()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'flint.types.nmod.nmod' object has no attribute 'parent'
>>> QQ.algebraic_field(sqrt(2)).one.parent()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'ANP' object has no attribute 'parent'
```
Some other domain elements do have `.parent()`:
```python
>>> ZZ['x'].one.parent()
ZZ[x]
>>> ZZ.frac_field('x').one.parent()
ZZ(x)
>>> ZZ_I.one.parent()
ZZ_I
```
I'm not sure how useful it is to define `.parent()` only for some types. We can't add `.parent()` to `int`, `mpz`, `fmpz`, `mpf` etc.

These are the places in the codebase that actually call `.parent()`:
```console
$ git grep '\.parent()' sympy
sympy/polys/agca/tests/test_extensions.py:    assert i.parent() is A
sympy/polys/domains/domain.py:            return self.convert_from(element, element.parent())
sympy/polys/domains/domainelement.py:        >>> p.parent()
sympy/polys/domains/tests/test_domains.py:        assert q.parent() == G
sympy/polys/solvers.py:                    K = val.parent()
```
The only one that is really useful is this one:
https://github.com/sympy/sympy/blob/817a52e6cfe157b5f461565f07fc47603885fc66/sympy/polys/domains/domain.py#L439-L440
That one is fine because it checks `isinstance(element, DomainElement)`. More generally though I think that `K.convert(val)` should not be used and it should always be `K1.convert_from(val, K2)`.

oscarbenjamin commented:
Part of the motivation for this PR is to be able to change the domain elements for RR and CC to Flint's `arf` type or perhaps even `float` and `complex`:
https://flintlib.org/doc/arf.html

Flint's `arf` is basically a faster version of mpmath's `mpf`. It is still not as fast as `float` but it is unclear to me yet how important it is that `RR` and `CC` use arbitrary precision. It is not currently used anywhere in the codebase as far as I can tell since `RR` and `CC` are just always 53 bits the same as float. There are other differences and so this is a discussion for elsewhere but at least `arf` could be used to speed up `RR`:
```python
In [1]: x1, x2 = RR(3), RR(5)

In [2]: %timeit x1 + x2
2.96 µs ± 123 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [3]: import flint

In [4]: a1, a2 = flint.arf(3), flint.arf(5)

In [5]: %timeit a1 + a2
154 ns ± 2.51 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [6]: f1, f2 = 3.0, 5.0

In [7]: %timeit f1 + f2
57.7 ns ± 7.14 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```
So `mpf` is about 20x slower than arf and 50x slower than float even when all three use the same 53 bit precision. This makes e.g. `DomainMatrix` over `RR` very slow.

sylee957 commented:
I agree that `.parent` may not be useful, and having that requires multiple inheritance which may make things fragile or slower, or even may not be possible with some wrapper libraries.
However, we may need to note about backward breaking changes here

oscarbenjamin commented:
> However, we may need to note about backward breaking changes here

I agree. There are two breaking changes here:

1. RealField and ComplexField no longer accept a tolerance argument and no longer have a tolerance attribute.
2. The elements of RR and CC no longer have a `.parent()` method.

We can't have `.parent()` for the elements if we want to be able to use `mpf` or `arf` for the elements so I think that is a necessary breaking change.

For the tolerance argument we could keep it and perhaps it could be just a no-op or that could be a warning about using it. I don't like what the argument is supposed to do but it basically sets `max_denom`. Then `max_denom` is used for converting `mpf` to `rational` to try to get nice rationals like `1/3` rather than the exact mpf value. More investigation would be needed to understand whether that can be removed but it certainly causes a handful of test failures to remove it right now.

oscarbenjamin commented:
I've added a breaking change not to the release notes about `.parent()`.

oscarbenjamin committed:
refactor(polys): Use mpmath mpf for RR and CC

Do not use the MPElement subclasses of mpmath's mpf and mpc and SymPy's
custom MPContext subclass. This means that RR and CC only use mpmath's
public interface.

This also fixes a bug where all RealField domains share a global
precision.

oscarbenjamin commented:
I don't know whether removing the `mpelements` module should be considered a compatibility break. In principle we can leave it there as an unused module with code that isn't needed any more. There are quite a few of those now:

- `sympy.polys.domains.gmpyfinitefield`
- `sympy.polys.domains.integerring`
- `sympy.polys.domains.rationalfield`
- `sympy.polys.domains.pythonfinitefield`
- `sympy.polys.domains.pythonintegerring`
- `sympy.polys.domains.pythonrationalfield`
- `sympy.polys.domains.pythonrational`

I've added it back for now.

oscarbenjamin commented:
Okay, I think this is ready.

sylee957 commented:
There can be some consequences of dropping `isinstance` with `DomainElement` for RR and CC.
For example, I see that `domain.convert` now follows the branches to use `sympify`, which may still be functional, but can be slower.

oscarbenjamin commented:
> For example, I see that `domain.convert` now follows the branches to use `sympify`, which may still be working, but can be slower to convert `RR` element to another `RR`

Yes, that is a good point. There should be an explicit check for mpf and mpc.

I will make a separate PR to remove all use of `convert` in favour of `convert_from` but for now I will add an explicit path here so that `convert` can recognise `mpf` and `mpc`.

oscarbenjamin committed:
fix(polys): detect mpf and mpc in Domaion.convert()

Handle mpf and mpc explicitly rather than falling back on sympify.

oscarbenjamin commented:
Thanks!
